torch/utils/data/_utils/worker.py

```diff
@@ -222,7 +223,17 @@ def _worker_loop(dataset_kind, dataset, index_queue, data_queue, done_event,
         torch.set_num_threads(1)
         seed = base_seed + worker_id
         random.seed(seed)
-        torch.manual_seed(seed)
+        generators = [o for o in gc.get_objects() if isinstance(o, torch.Generator)]
+        for generator in generators:
+            if generator is torch.random.default_generator:
+                torch.manual_seed(seed)
+            else:
+                # Note that we can't use generator.manual_seed(seed) here, otherwise the RNG of the generator
+                # would be the same as the global one.
+                # We also need to create a generator_seed that depends on the current generator state, otherwise
+                # all Generator instances within a given worker would have the same RNG.
+                generator_seed = torch.empty((), dtype=torch.int64).random_(generator=generator).item() + seed
+                generator.manual_seed(generator_seed)
```

Comment:
@albanD regarding https://github.com/pytorch/pytorch/pull/107034#discussion_r1291699664

> Why not use generator.initial_seed() ?

I'm relying on the same logic that is used to generate the `base_seed` (used to seed the default Generator just above): 
https://github.com/pytorch/pytorch/blob/937cd3742b959c9d847f2770530c7e393092ddc7/torch/utils/data/dataloader.py#L601C8-L601C8

Using `initial_seed()` would reproduce the same RNG that existed before the `fork()`. I'm not sure it matters in practice?

Reply:
Ho I meant `new_seed = generator.initial_seed() + seed`

Reply:
Ho I didn't knew that this was how the seed is gathered.
Keeping both in sync sounds best, sorry for the confusion. Maybe a small helper to avoid duplicating this code would be the best here.